File: web-app/src/lib/domain/exercise-catalog.ts
import type { ConditioningScoringType } from "./types";

// ---------------------------------------------------------------------------
// Weightlifting Exercise Catalog
// ---------------------------------------------------------------------------

export type WeightliftingCategory =
  | "Squat"
  | "Hip Hinge"
  | "Press"
  | "Pull"
  | "Carry"
  | "Olympic Weightlifting";

export interface WeightliftingEntry {
  id: string;
  category: WeightliftingCategory;
}

export const WEIGHTLIFTING_EXERCISES: WeightliftingEntry[] = [
  // Squat
  { id: "Back Squat",            category: "Squat" },
  { id: "Front Squat",           category: "Squat" },
  { id: "Safety Bar Squat",      category: "Squat" },
  { id: "Box Squat",             category: "Squat" },
  { id: "Pause Squat",           category: "Squat" },
  { id: "Goblet Squat",          category: "Squat" },
  // Hip Hinge
  { id: "Conventional Deadlift (No Straps)",   category: "Hip Hinge" },
  { id: "Conventional Deadlift (With Straps)", category: "Hip Hinge" },
  { id: "Romanian Deadlift (No Straps)",       category: "Hip Hinge" },
  { id: "Romanian Deadlift (With Straps)",     category: "Hip Hinge" },
  { id: "Sumo Deadlift (No Straps)",           category: "Hip Hinge" },
  { id: "Sumo Deadlift (With Straps)",         category: "Hip Hinge" },
  { id: "Trap Bar Deadlift (No Straps)",       category: "Hip Hinge" },
  { id: "Trap Bar Deadlift (With Straps)",     category: "Hip Hinge" },
  { id: "Good Morning",          category: "Hip Hinge" },
  { id: "Hip Thrust",            category: "Hip Hinge" },
  // Press
  { id: "Flat Barbell Bench Press",    category: "Press" },
  { id: "Incline Barbell Bench Press", category: "Press" },
  { id: "Strict Press",                category: "Press" },
  { id: "Push Press",                  category: "Press" },
  { id: "Dumbbell Bench Press",        category: "Press" },
  { id: "Dumbbell Overhead Press",     category: "Press" },
  // Pull
  { id: "Barbell Row",       category: "Pull" },
  { id: "Pendlay Row",       category: "Pull" },
  { id: "Pull-Up",           category: "Pull" },
  { id: "Weighted Pull-Up",  category: "Pull" },
  { id: "Lat Pulldown",      category: "Pull" },
  { id: "Cable Row",         category: "Pull" },
  // Carry
  { id: "Farmers Carry",   category: "Carry" },
  { id: "Suitcase Carry",  category: "Carry" },
  // Olympic Weightlifting
  { id: "Squat Snatch",    category: "Olympic Weightlifting" },
  { id: "Squat Clean",     category: "Olympic Weightlifting" },
  { id: "Power Clean",     category: "Olympic Weightlifting" },
  { id: "Power Snatch",    category: "Olympic Weightlifting" },
  { id: "Hang Clean",      category: "Olympic Weightlifting" },
  { id: "Hang Snatch",     category: "Olympic Weightlifting" },
  { id: "Split Jerk",      category: "Olympic Weightlifting" },
  { id: "Push Jerk",       category: "Olympic Weightlifting" },
  { id: "Clean and Jerk",  category: "Olympic Weightlifting" },
];

const _weightliftingIDs = new Set(WEIGHTLIFTING_EXERCISES.map((e) => e.id));

export function isWeightliftingExercise(id: string): boolean {
  return _weightliftingIDs.has(id);
}

// ---------------------------------------------------------------------------
// Conditioning Exercise Catalog
// ---------------------------------------------------------------------------

export interface ConditioningEntry {
  id: string;
  defaultScoringType: ConditioningScoringType;
}

export const CONDITIONING_EXERCISES: ConditioningEntry[] = [
  // Reps-based
  { id: "Wall Ball",               defaultScoringType: "reps" },
  { id: "Box Jump",                defaultScoringType: "reps" },
  { id: "Burpee",                  defaultScoringType: "reps" },
  { id: "Kettlebell Swing",        defaultScoringType: "reps" },
  { id: "Double Under",            defaultScoringType: "reps" },
  { id: "Pull-Up (Kipping)",       defaultScoringType: "reps" },
  { id: "Toes-to-Bar",             defaultScoringType: "reps" },
  { id: "Muscle-Up",               defaultScoringType: "reps" },
  { id: "Push-Up",                 defaultScoringType: "reps" },
  { id: "Sit-Up",                  defaultScoringType: "reps" },
  { id: "Air Squat",               defaultScoringType: "reps" },
  { id: "Thruster",                defaultScoringType: "reps" },
  { id: "Rowing (Calories)",       defaultScoringType: "reps" },
  { id: "Assault Bike (Calories)", defaultScoringType: "reps" },
  // Time-based
  { id: "400m Run",        defaultScoringType: "time" },
  { id: "800m Run",        defaultScoringType: "time" },
  { id: "1-Mile Run",      defaultScoringType: "time" },
  { id: "5K Run",          defaultScoringType: "time" },
  { id: "500m Row",        defaultScoringType: "time" },
  { id: "2K Row",          defaultScoringType: "time" },
  { id: "1K Assault Bike", defaultScoringType: "time" },
];

const _conditioningMap = new Map(
  CONDITIONING_EXERCISES.map((e) => [e.id, e.defaultScoringType])
);

export function isConditioningExercise(id: string): boolean {
  return _conditioningMap.has(id);
}

export function conditioningScoringType(
  id: string
): ConditioningScoringType | undefined {
  return _conditioningMap.get(id);
}

/**
 * Formats a conditioning value for display.
 * - time:  value in seconds → "M:SS"
 * - reps:  count → "N reps" (singular: "1 rep")
 */
export function formatConditioningValue(
  value: number,
  type: ConditioningScoringType
): string {
  if (type === "time") {
    const totalSeconds = Math.floor(value);
    const minutes = Math.floor(totalSeconds / 60);
    const seconds = totalSeconds % 60;
    return `${minutes}:${String(seconds).padStart(2, "0")}`;
  }
  const count = Math.floor(value);
  return count === 1 ? "1 rep" : `${count} reps`;
}

/**
 * Returns true when `newValue` is a better score than `existingValue`.
 * If `existingValue` is undefined the new score is always better.
 * - time: lower is better
 * - reps: higher is better
 */
export function isBetterConditioningScore(
  newValue: number,
  existingValue: number | undefined,
  type: ConditioningScoringType
): boolean {
  if (existingValue === undefined) return true;
  return type === "time" ? newValue < existingValue : newValue > existingValue;
}

File: android/core/src/main/java/com/sundeefundee/core/domain/exercise/ExerciseCatalog.kt
package com.sundeefundee.core.domain.exercise

// MARK: - Weightlifting Category

enum class WeightliftingCategory {
    SQUAT,
    HIP_HINGE,
    PRESS,
    PULL,
    CARRY,
    OLYMPIC_WEIGHTLIFTING,
}

data class WeightliftingEntry(
    val id: String,
    val category: WeightliftingCategory,
)

object WeightliftingCatalog {
    val all: List<WeightliftingEntry> = listOf(
        // Squat
        WeightliftingEntry("Back Squat", WeightliftingCategory.SQUAT),
        WeightliftingEntry("Front Squat", WeightliftingCategory.SQUAT),
        WeightliftingEntry("Safety Bar Squat", WeightliftingCategory.SQUAT),
        WeightliftingEntry("Box Squat", WeightliftingCategory.SQUAT),
        WeightliftingEntry("Pause Squat", WeightliftingCategory.SQUAT),
        WeightliftingEntry("Goblet Squat", WeightliftingCategory.SQUAT),
        WeightliftingEntry("Leg Press", WeightliftingCategory.SQUAT),
        WeightliftingEntry("Hack Squat", WeightliftingCategory.SQUAT),
        WeightliftingEntry("Leg Extension", WeightliftingCategory.SQUAT),
        // Hip Hinge
        WeightliftingEntry("Conventional Deadlift (No Straps)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Conventional Deadlift (With Straps)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Romanian Deadlift (No Straps)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Romanian Deadlift (With Straps)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Sumo Deadlift (No Straps)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Sumo Deadlift (With Straps)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Trap Bar Deadlift (No Straps)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Trap Bar Deadlift (With Straps)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Good Morning", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Hip Thrust", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Leg Curl (Lying)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Leg Curl (Seated)", WeightliftingCategory.HIP_HINGE),
        WeightliftingEntry("Glute Ham Raise", WeightliftingCategory.HIP_HINGE),
        // Press
        WeightliftingEntry("Flat Barbell Bench Press", WeightliftingCategory.PRESS),
        WeightliftingEntry("Incline Barbell Bench Press", WeightliftingCategory.PRESS),
        WeightliftingEntry("Decline Barbell Bench Press", WeightliftingCategory.PRESS),
        WeightliftingEntry("Strict Press", WeightliftingCategory.PRESS),
        WeightliftingEntry("Push Press", WeightliftingCategory.PRESS),
        WeightliftingEntry("Dumbbell Bench Press", WeightliftingCategory.PRESS),
        WeightliftingEntry("Dumbbell Incline Press", WeightliftingCategory.PRESS),
        WeightliftingEntry("Dumbbell Overhead Press", WeightliftingCategory.PRESS),
        WeightliftingEntry("Dips (Weighted)", WeightliftingCategory.PRESS),
        WeightliftingEntry("Close Grip Bench Press", WeightliftingCategory.PRESS),
        // Pull
        WeightliftingEntry("Barbell Row", WeightliftingCategory.PULL),
        WeightliftingEntry("Pendlay Row", WeightliftingCategory.PULL),
        WeightliftingEntry("Pull-Up", WeightliftingCategory.PULL),
        WeightliftingEntry("Weighted Pull-Up", WeightliftingCategory.PULL),
        WeightliftingEntry("Lat Pulldown", WeightliftingCategory.PULL),
        WeightliftingEntry("Cable Row", WeightliftingCategory.PULL),
        WeightliftingEntry("Dumbbell Row", WeightliftingCategory.PULL),
        WeightliftingEntry("Face Pull", WeightliftingCategory.PULL),
        WeightliftingEntry("Bicep Curl (Barbell)", WeightliftingCategory.PULL),
        WeightliftingEntry("Bicep Curl (Dumbbell)", WeightliftingCategory.PULL),
        WeightliftingEntry("Hammer Curl", WeightliftingCategory.PULL),
        WeightliftingEntry("Upright Row", WeightliftingCategory.PULL),
        // Carry
        WeightliftingEntry("Farmers Carry", WeightliftingCategory.CARRY),
        WeightliftingEntry("Suitcase Carry", WeightliftingCategory.CARRY),
        WeightliftingEntry("Zercher Carry", WeightliftingCategory.CARRY),
        WeightliftingEntry("Yoke Walk", WeightliftingCategory.CARRY),
        // Olympic Weightlifting
        WeightliftingEntry("Squat Snatch", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Squat Clean", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Power Clean", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Power Snatch", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Hang Clean", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Hang Snatch", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Split Jerk", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Push Jerk", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Clean and Jerk", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Muscle Snatch", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
        WeightliftingEntry("Muscle Clean", WeightliftingCategory.OLYMPIC_WEIGHTLIFTING),
    )

    private val idSet: Set<String> = all.map { it.id }.toSet()

    fun contains(id: String): Boolean = idSet.contains(id)
}

// MARK: - Conditioning Category

enum class ConditioningScoringType {
    TIME,
    REPS,
}

data class ConditioningEntry(
    val id: String,
    val defaultScoringType: ConditioningScoringType,
)

object ConditioningCatalog {
    val all: List<ConditioningEntry> = listOf(
        // Reps-based
        ConditioningEntry("Wall Ball", ConditioningScoringType.REPS),
        ConditioningEntry("Box Jump", ConditioningScoringType.REPS),
        ConditioningEntry("Burpee", ConditioningScoringType.REPS),
        ConditioningEntry("Kettlebell Swing", ConditioningScoringType.REPS),
        ConditioningEntry("Double Under", ConditioningScoringType.REPS),
        ConditioningEntry("Pull-Up (Kipping)", ConditioningScoringType.REPS),
        ConditioningEntry("Toes-to-Bar", ConditioningScoringType.REPS),
        ConditioningEntry("Muscle-Up", ConditioningScoringType.REPS),
        ConditioningEntry("Push-Up", ConditioningScoringType.REPS),
        ConditioningEntry("Sit-Up", ConditioningScoringType.REPS),
        ConditioningEntry("Air Squat", ConditioningScoringType.REPS),
        ConditioningEntry("Thruster", ConditioningScoringType.REPS),
        ConditioningEntry("Rowing (Calories)", ConditioningScoringType.REPS),
        ConditioningEntry("Assault Bike (Calories)", ConditioningScoringType.REPS),
        ConditioningEntry("SkiErg (Calories)", ConditioningScoringType.REPS),
        ConditioningEntry("Box Step-up", ConditioningScoringType.REPS),
        ConditioningEntry("Lunges (Alternating)", ConditioningScoringType.REPS),
        ConditioningEntry("Wall Walk", ConditioningScoringType.REPS),
        ConditioningEntry("Handstand Push-up", ConditioningScoringType.REPS),
        ConditioningEntry("GHD Sit-up", ConditioningScoringType.REPS),
        ConditioningEntry("V-up", ConditioningScoringType.REPS),
        // Time-based
        ConditioningEntry("400m Run", ConditioningScoringType.TIME),
        ConditioningEntry("800m Run", ConditioningScoringType.TIME),
        ConditioningEntry("1-Mile Run", ConditioningScoringType.TIME),
        ConditioningEntry("5K Run", ConditioningScoringType.TIME),
        ConditioningEntry("500m Row", ConditioningScoringType.TIME),
        ConditioningEntry("2K Row", ConditioningScoringType.TIME),
        ConditioningEntry("1K Assault Bike", ConditioningScoringType.TIME),
        ConditioningEntry("2K SkiErg", ConditioningScoringType.TIME),
        ConditioningEntry("Plank Hold", ConditioningScoringType.TIME),
        ConditioningEntry("L-Sit Hold", ConditioningScoringType.TIME),
    )
}

// MARK: - Helpers

fun isWeightliftingExercise(id: String): Boolean = WeightliftingCatalog.contains(id)
